### DISCUSSION TITLE
Qt on Unikraft

### DISCUSSION THREAD

**Discussion by Rochet2** [unikraft/unikraft#353] [Q&A]
I am trying to look into if Qt framework could be used with unikraft.
Any thoughts on Qt support?

For now Im testing a basic console application that only uses the Qt core and has no GUI to get something running.
There are some basic problems in getting anything to compile due to some ongoing issues in unikraft.
One issue I have is large file support that I cant seem to find how to disable on Qt or enable on unikraft.
Any help with that is appreciated.

Here is a list of undefined references, however some are undefined because I am unable to link all of Qt libraries with unikraft at this time.
There might be some missing calls also at this moment in unikraft. Would be interested to hear about what those are and if some could be solved with existing libraries that I have missed.

<details>

```
undefined references to `__errno_location' follow
undefined references to `std::__throw_system_error(int)' follow
undefined reference to `backtrace'
undefined reference to `backtrace_symbols'
undefined reference to `__ctype_b_loc'
undefined reference to `__errno_location'
undefined reference to `eventfd_read'
undefined reference to `eventfd_write'
undefined reference to `__fprintf_chk'
undefined reference to `fseeko64'
undefined reference to `ftello64'
undefined reference to `getgrgid_r'
undefined reference to `g_main_context_default'
undefined reference to `g_main_context_iteration'
undefined reference to `g_main_context_new'
undefined reference to `g_main_context_pending'
undefined reference to `g_main_context_pop_thread_default'
undefined reference to `g_main_context_push_thread_default'
undefined reference to `g_main_context_ref'
undefined reference to `g_main_context_unref'
undefined reference to `g_main_context_wakeup'
undefined reference to `g_source_add_poll'
undefined reference to `g_source_attach'
undefined reference to `g_source_destroy'
undefined reference to `g_source_new'
undefined reference to `g_source_remove_poll'
undefined reference to `g_source_set_can_recurse'
undefined reference to `g_source_unref'
undefined reference to `__memcpy_chk'
undefined reference to `__memset_chk'
undefined reference to `mmap64'
undefined reference to `posix_madvise'
undefined reference to `__printf_chk'
undefined reference to `__snprintf_chk'
undefined reference to `__sprintf_chk'
undefined reference to `std::__atomic_futex_unsigned_base::_M_futex_wait_until(unsigned int*, unsigned int, bool, std::chrono::duration<long, std::ratio<1l, 1l> >, std::chrono::duration<long, std::ratio<1l, 1000000000l> >)'
undefined reference to `std::ctype<char> const& std::use_facet<std::ctype<char> >(std::locale const&)'
undefined reference to `std::__detail::_Prime_rehash_policy::_M_need_rehash(unsigned long, unsigned long, unsigned long) const'
undefined reference to `std::__detail::_Prime_rehash_policy::_M_next_bkt(unsigned long) const'
undefined reference to `stderr'
undefined reference to `std::__exception_ptr::exception_ptr::~exception_ptr()'
undefined reference to `std::__exception_ptr::exception_ptr::exception_ptr(std::__exception_ptr::exception_ptr const&)'
undefined reference to `std::__exception_ptr::operator==(std::__exception_ptr::exception_ptr const&, std::__exception_ptr::exception_ptr const&)'
undefined reference to `std::locale::classic()'
undefined reference to `std::pmr::get_default_resource()'
undefined reference to `std::pmr::monotonic_buffer_resource::_M_new_buffer(unsigned long, unsigned long)'
undefined reference to `std::pmr::monotonic_buffer_resource::~monotonic_buffer_resource()'
undefined reference to `std::rethrow_exception(std::__exception_ptr::exception_ptr)'
undefined reference to `std::__throw_bad_function_call()'
undefined reference to `std::__throw_future_error(int)'
undefined reference to `std::__throw_length_error(char const*)'
undefined reference to `std::__throw_system_error(int)'
undefined reference to `timezone'
undefined reference to `typeinfo for __cxxabiv1::__forced_unwind'
undefined reference to `tzname'
undefined reference to `vtable for std::pmr::monotonic_buffer_resource'
```

</details>

**Comment by vladandrew** [comment]
Hi Rocket2,

Qt should be interesting to port on Unikraft, how did you chose it?

You seems to be missing some unported external libraries(e.g. pcre2), libc functions(e.g. mmap64, timezone) and some  C++ symbols. I think the first of action would be to port pcre2 on Unikraft, stub the missing libc functions and a find where the C++ symbols are defined(e.g. std::__throw_system_error). I recommend cscope to aid you in this task. Since yhe number of symbols is rather low,  you shpuld be able to fix it quickly.

**Reply by Rochet2** [reply]
I now updated the list of missing symbols as I have been able to get forward slightly.

> how did you chose it?

It has support for graphical and console applications. Its suitable for embedded as well as desktop, and web through Wasm. It has a large enterprise backing. Existing applications could potentially be run with it supported.

I think one important part is also its modularity that enables customization and specialization. Qt is split into large modules that each implement a functionality, including network, media, GUI and SQL support. They can be disabled if unused.

> first of action would be to port pcre2 on Unikraft

Qt has an internal version of pcre2, so I can just use that and porting isn't necessary then. At least for now.

> stub the missing libc functions
> find where the C++ symbols are defined

Hmm, need to look into how this is done.

**Comment by razvand** [comment]
@Rochet2, since we are already (viciously) using [the Unikraft Discord server](https://bit.ly/UnikraftDiscord) for discussions, meetings, brainstorming and jokes (mostly bad, we do what we can), we will close the Discussions tab. We will do this on Friday, March 31, 2023.

Please move any discussions or questions on [the Unikraft Discord server](https://bit.ly/UnikraftDiscord). See you all there! :handshake:
